### ISSUE TITLE
Pooling: one document vector from N chunk vectors

### ISSUE THREAD

**Issue by charitarthchugh** [charitarthchugh/paperscale#25]
## Question

Standing decision 3 requires one pooled document vector per document, and the common case is
a single chunk where pooling must be the identity.

- **Weighting.** Uniform mean, token-weighted mean, or length-weighted? Token weighting needs
  a token count per chunk, which #23 may or may not make cheap.
- **Normalization order.** L2-normalize each chunk before pooling and renormalize after, or
  pool raw and normalize once? These give different answers: normalizing first makes a chunk
  count for how much text it represents rather than how large its raw norm happens to be.
- **Is a mean even right?** For a 300-page deposition, the mean of 40 chunk vectors may be
  close to meaningless. Alternatives: store no document vector for very long documents, or
  use a different reduction. But standing decision 3 requires a uniform shape, so "no vector"
  would have to mean a defined sentinel, not an absent array.
- **Verify the identity case:** with one chunk, the pooled vector must equal the chunk vector
  exactly, not approximately.

Depends on #24, since the pooling weights are a property of how chunks were cut.

**Comment by charitarthchugh** [comment]
Context from [Chunking strategy when a document exceeds the discovered limit](https://github.com/charitarthchugh/paperscale/issues/24), now closed:

Chunking is **greedy page packing with no overlap**, so chunks are **unequal in size** — a tail chunk may hold one page against a predecessor's forty. Two things follow for this ticket:

1. **Mean pooling over chunk vectors is length-biased.** An unweighted mean gives that one-page tail chunk the same say in the document vector as the forty-page chunk before it. Whether to weight by `token_count` is a decision this ticket now has to make; the field is recorded per chunk either way.
2. **No overlap means the chunks partition the text exactly once**, so weighting by `token_count` is a genuine length weighting with no double-counted spans — that was part of why overlap was ruled out.

The per-chunk fields available to pool over: `start_char`, `end_char`, `first_page`, `last_page`, `chunk_index`, `n_chunks`, `token_count`, `is_partial_page`.

**Comment by charitarthchugh** [comment]
Second note, from [Is MRL truncation offered, and how does an adapter express validity?](https://github.com/charitarthchugh/paperscale/issues/34), now closed:

Pooling operates on the **truncated, re-normalized chunk vectors** — not on native ones. `--embed-dim` (default 768) is applied client-side to each chunk vector, which is then re-normalized; only then does this ticket's reduction run, and the resulting document vector is re-normalized in turn.

The reason the order was fixed that way is a property this ticket inherits: **a consumer holding the sink can recompute the document vector from the stored chunk vectors and get the same answer.** Pooling native vectors and truncating afterwards would break that, because the document vector would depend on dimensions that were never written. Whatever weighting this ticket chooses should preserve the property — it needs to be computable from `chunk_vector` and the recorded per-chunk fields alone.

**Comment by charitarthchugh** [comment]
## Resolution

**Token-weighted mean of the stored chunk vectors, computed for every document, with an explicit
short-circuit for the single-chunk case.**

### Two of this ticket's four bullets were closed upstream

**Normalization order** was fixed by [#34](https://github.com/charitarthchugh/paperscale/issues/34),
and one of the options this ticket listed is no longer reachable. Chunk vectors are sliced to
`stored_dim` and re-normalized *before* pooling, so every stored chunk vector is unit length by
construction. "Pool raw and normalize once" would need the unnormalized pooled vectors
(`use_activation: false` on the request), and the raw norms are neither stored nor recoverable —
so that route breaks #34's reproducibility property outright.

**Token counts** were made cheap and exact by
[#24](https://github.com/charitarthchugh/paperscale/issues/24). `token_count` is one of the eight
recorded per-chunk fields, taken from `/v1/tokenize`. The ticket's "#23 may or may not make this
cheap" is resolved.

### The decisions

**Weighting: token-weighted.** Each chunk vector contributes in proportion to its `token_count`.

The deciding property is **invariance to the cut**. The chunk budget is an implementation detail
— it moves when the model's validated context length moves, and #24's oversized-page path can
move it again within a single document. A token-weighted mean stays approximately put across
those changes; a uniform mean swings hard. Greedy packing makes this concrete: a two-chunk
document can be 40 pages plus a 1-page tail, and a uniform mean hands that tail page half the
document vector.

Recorded as a knowing trade rather than a free win: because chunk vectors are unit length, token
weighting asserts that text volume equals importance. In a legal document a one-page cover letter
may be more discriminative than a forty-page exhibit. That is a claim about this corpus that
cannot be tested inside this effort, and "proportional to content" is the neutral default — it is
also what a single-shot embedding of the whole document would approximate, which is the thing
chunking stands in for.

**Every document gets a document vector. No sentinel, no threshold.** The dilution concern is
real — averaging forty diverse chunk vectors does drift toward the corpus centroid — but a
sentinel needs a chunk-count threshold, and no measurement available in this effort could
justify one. The map puts embedding-quality evaluation out of scope, and `CONTEXT.md` assigns
exactly this judgement to the **Consumer**. Picking a cutoff here would be inventing a number.

The safety net already exists: chunk vectors are stored, so a consumer that finds the document
vector too diluted builds its own reduction from the primary data. `n_chunks` is recorded, so
that judgement is available without reading the vectors at all. The document vector is a
convenience; the chunk vectors are the output.

**The single-chunk case is a copy, not an average.** With `n_chunks == 1` the document vector
must equal the chunk vector exactly. It does not fall out of the arithmetic: a weighted mean of
one element is then re-normalized, and the divisor is `1 ± ε` rather than `1`, so the low bits
move. paperscale short-circuits and copies. A test asserts bit-equality, not approximate
equality.

### The algorithm

```
chunk_vector[i] = normalize(slice(server_vector[i], stored_dim))   # from #34

if n_chunks == 1:
    document_vector = copy(chunk_vector[0])            # exact; no arithmetic runs
else:
    w = [chunk.token_count for chunk in chunks]
    document_vector = normalize(sum(w[i] * chunk_vector[i] for i in range(n_chunks)))
```

Note the missing division. Dividing the weighted sum by `sum(w)` is a no-op, because the
normalize that follows discards any positive scale. Leaving it out removes a step and one source
of float error.

**Edge case, defined rather than left open:** if `sum(w) == 0`, fall back to uniform weights.
This is unreachable in practice — a document with more than one chunk overflowed, so it has
tokens — but the alternative is a division by zero in a path nobody tests. Separately, a weighted
sum with zero norm would require the chunk vectors to cancel exactly; treat that as an error
rather than writing `NaN` into a sink.

### Properties

- **The document vector is reproducible from the sink.** A consumer holding `chunk_vector` and
  `token_count` recomputes it and gets the same answer. #34 asked this ticket to preserve that
  property; token weighting preserves it, because the weights are recorded per chunk.
- **Invariance to the cut is approximate, not exact.** A chunk vector covering forty concatenated
  pages is not the mean of forty single-page vectors, so re-chunking the same document at a
  different budget shifts the document vector slightly. The claim is that it shifts far less than
  under uniform weighting, not that it is fixed.
- **The identity case is exact by construction**, not by tolerance.

### What this hands to other tickets

- **#26 / #27 (sinks):** the document vector has the same width as the chunk vectors
  (`stored_dim`). Both sinks must keep `token_count` readable alongside the chunk vectors, since
  the recomputability guarantee above depends on it — it is not merely diagnostic.
- **#26** is unblocked by this resolution.

No new tickets.

**Comment by charitarthchugh** [comment]
## Correction — the ranking in this resolution was retracted

This resolution contains, in the argument for giving every Document a vector with no sentinel:

> The safety net already exists: chunk vectors are stored, so a consumer that finds the document vector too diluted builds its own reduction from **the primary data**. `n_chunks` is recorded, so that judgement is available without reading the vectors at all. **The document vector is a convenience; the chunk vectors are the output.**

**That ranking is retracted.** It was retracted at the map level (standing decision 3, amended after [#36](https://github.com/charitarthchugh/paperscale/issues/36)) and on [#31](https://github.com/charitarthchugh/paperscale/issues/31). [#39](https://github.com/charitarthchugh/paperscale/issues/39) then swept the record for exactly this sentence — and **targeted #36 only, missing this ticket**. So it has stayed live here.

**Why it is wrong.** paperscale serves two Consumers and they read different artifacts. **RAG** reads Chunk vectors — ranking is max-over-Chunks, which is query-*dependent*, so a stored Document vector cannot compete. **Classification** reads Document vectors — `sklearn`/`xgboost`/torch want `X` of shape `(n_documents, stored_dim)` against one label per Document, and with no query the reduction cannot be deferred to read time. Neither artifact is primary; neither is a convenience over the other. Both Sinks carry both.

**Everything else in this resolution stands**, including the part the retracted sentence was supporting: every Document gets a Document vector, no sentinel and no chunk-count threshold, because picking a cutoff would be inventing a number and `CONTEXT.md` assigns that judgement to the **Consumer**. The token-weighted mean, the exact single-Chunk copy, and the recomputability guarantee resting on a readable `token_count` are all unaffected.

*Found while reviewing the design document written for #31, which is what #39's pass was meant to protect. Recorded here so the sweep is actually complete.*
